### PR TITLE
Use DigestIO to reduce I/O when parsing compact index file digests

### PR DESCRIPTION
**Pull request by martinemde**
## What was the end-user or developer problem that led to this PR?

When reviewing #6971, @simi asked about moving the MD5 digest code to its own helper. When looking at the code it seemed like we could calculate the digest as the files were moved and saved. I decided to see if it would work.

## What is your fix for the problem, implemented in this PR?

This builds on #6971 using `Gem::Package::DigestIO` to read the compact index file digests while the other file movement operations are being performed, saving a few extra file reads.

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [X] Write code to solve the problem
- [X] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by martinemde** [comment]
I switched the base to @segiddins branch so that we could separate his changes from what I just wrote here so it's less confusing.

**Comment by segiddins** [comment]
What's the comparison of allocations here?

**Comment by martinemde** [comment]
@segiddins Ok, this was closer than I thought. The memory usage is barely different, but DigestIO is slightly higher (I even tried making our own DigestIO that doesn't take a hash). Overall, the usage here pales in comparison to the rest of the app.

Tested by running `bundle lock --update --bundler` on a client app I have locally.

```
==> memprof.before.txt <==
Total allocated: 398.62 MB (5878673 objects)
Total retained:  107.41 MB (1469380 objects)
==> memprof.custom_digest_io.txt <==
Total allocated: 370.14 MB (5634274 objects)
Total retained:  104.96 MB (1412709 objects)
==> memprof.rubygems_digestio.txt <==
Total allocated: 370.15 MB (5634247 objects)
Total retained:  104.96 MB (1412706 objects)
==> memprof.segiddins.txt <==
Total allocated: 369.68 MB (5629483 objects)
Total retained:  104.90 MB (1412106 objects)
```

Not much, still better than before, but your version does allocate less memory. 

But that's not the whole story. I couldn't believe there was no advantage to avoiding extra reads of the same file. The advantage is that it's much faster.

```
Warming up --------------------------------------
           digest_io     1.781k i/100ms
    copy then digest     1.193k i/100ms
Calculating -------------------------------------
           digest_io     17.251k (± 8.9%) i/s -     85.488k in   5.004996s
    copy then digest     11.590k (± 6.7%) i/s -     58.457k in   5.072384s

Comparison:
           digest_io:    17251.3 i/s
    copy then digest:    11590.4 i/s - 1.49x  slower
```

<details>
  <summary>benchmark.rb</summary>
    # frozen_string_literal: true

    require 'digest'
    require 'benchmark/ips'

    class DigestIO
      def initialize(io, digest)
        @io = io
        @digest = digest
      end

      def write(data)
        @digest << data
        @io.write(data)
      end
    end

    def copy_file(source, dest, digest)
      File.open(source, "r") do |s|
        File.open(dest, "wb", s.stat.mode) do |f|
          out = digest ? DigestIO.new(f, digest) : f
          IO.copy_stream(s, out)
        end
      end
    end

    def calcdigest(path)
      File.open(path, "rb") do |f|
        digest = Digest::MD5.new
        buf = String.new(:capacity => 16_384, :encoding => Encoding::BINARY)
        digest << buf while f.read(16_384, buf)
        digest.hexdigest
      end
    end

    source = 'tmp/1/gems/remote1/specs.4.8'
    dest = 'tmp/deleteme'

    Benchmark.ips do |x|
      x.report("digest_io") do
        digest = Digest::MD5.new
        copy_file(source, dest, digest)
        digest.hexdigest
      end

      x.report("copy then digest") do
        copy_file(source, dest, nil)
        calcdigest(dest)
      end

      x.compare!
    end
</details>

File: bundler/lib/bundler/compact_index_client/updater.rb [hunk 1]
```diff
@@ -80,38 +84,30 @@ def update(local_path, remote_path, retrying = nil)
...
+
+      def digest_io(io, digests)
+        Gem::Package::DigestIO.new(io, digests)
       end
 
       def slice_body(body, range)
         body.byteslice(range)
       end
 
       def checksum_for_file(path)
```
Comment (segiddins) [inline bundler/lib/bundler/compact_index_client/updater.rb:99]: can this method be removed now that `etag_for` uses the output from `digest_io` ?

File: bundler/lib/bundler/compact_index_client/updater.rb [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require_relative "../vendored_fileutils"
+require "rubygems/package"
```
Comment (segiddins) [inline bundler/lib/bundler/compact_index_client/updater.rb:4]: can we instead use `gem/package/digest_io` ?

**Review by segiddins — APPROVED** [review]
will defer to your judgement on which version you prefer, @martinemde

File: bundler/lib/bundler/compact_index_client/updater.rb (hunk 2 above)
Comment (martinemde) [reply bundler/lib/bundler/compact_index_client/updater.rb:4]: unfortunately that causes the error `NameError: uninitialized constant Gem::Package`

**Comment by martinemde** [comment]
I'm continuing this in #7122 and closing this version. This only really makes sense when merged with the opaque etag work.

**Pull request closed without merging**